### PR TITLE
[many  ports] Add X windows part 11/N

### PR DESCRIPTION
**Pull request by Neumann-A**

### CODE REVIEW

File: ports/libxaw/getpagesize.patch [hunk 1]
```diff
@@ -0,0 +1,14 @@
+diff --git a/src/OS.c b/src/OS.c
+index 4c69005..cd0c517 100644
+--- a/src/OS.c
++++ b/src/OS.c
+@@ -37,6 +37,9 @@ _XawGetPageSize(void)
+ #endif
+ 
+ #ifdef HAVE_GETPAGESIZE
++# if defined(_WIN32)
++extern int getpagesize(void);
```
Comment (Neumann-A) [inline ports/libxaw/getpagesize.patch:10]: since this links I assume it is just missing the correct header on windows. Anybody know which one ?

Comment (Neumann-A) [reply ports/libxaw/getpagesize.patch]: ok this code path is never taken

File: ports/libx11/thread-safe-init.patch [hunk 2]
```diff
@@ -0,0 +1,33 @@
+diff --git a/src/globals.c b/src/globals.c
+index ea3fe6e11..3a433f249 100644
+--- a/src/globals.c
++++ b/src/globals.c
+@@ -98,9 +98,26 @@ int			XTestFakeAckType   = 1;
+ #endif
+ 
+ #ifdef USE_THREAD_SAFETY_CONSTRUCTOR
```
Comment (Neumann-A) [inline ports/libx11/thread-safe-init.patch:8]: I could deactivate this via a configure flag for windows instead of patching the below in but since the idea is to locate apps which use x incorrectly I thought it would be good to have.

File: ports/libxt/globals.patch [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
++# if defined(_MSC_VER) && defined(XT_DLL_EXPORTS)
++#  ifdef XT_BUILD
++#   define XT_EXTERN_API extern __declspec(dllexport)
++#  else
++#   define XT_EXTERN_API extern __declspec(dllimport)
++#  endif
++# else
++#  define XT_EXTERN_API extern
++# endif
++#endif
```
Comment (Neumann-A) [inline ports/libxt/globals.patch:32]: I could remove this due to:
```
if(VCPKG_TARGET_IS_WINDOWS)
    vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
    # Downstream uses &widgetClassRec in a const context which doesn't work 
    # if this is a dynamic library since the memory adress is only known at runtime
endif()
```
but since the work is already done and works I thought about leaving it in. Maybe there is some trick to make &widgetClassRec in const context work, which I dont know about.

File: versions/l-/libx11.json [hunk 4]
```diff
@@ -1,5 +1,15 @@
...
+    {
+      "git-tree": "bd487d761061c63f8250aaf18052cc6d25143a5c",
+      "version": "1.8.1",
+      "port-version": 0
+    },
+    {
+      "git-tree": "05282e1e4d501c80dc305b386ce7b814296a66da",
+      "version": "1.7.3.1",
+      "port-version": 2
+    },
```
Comment (JonLiu1993) [inline versions/l-/libx11.json:12]: @Neumann-A , Thanks for your pr, you should remove this code.

File: scripts/ci.baseline.txt [hunk 5]
```diff
@@ -550,6 +550,9 @@ libwandio:x64-windows-static-md=fail
 libwandio:x64-uwp=fail
 libwandio:arm64-windows=fail
 libwandio:arm-uwp=fail
+libxaw:x64-windows-static=skip
+# clang-cl ICEd
+libxt:arm64-windows=fail
```
Comment (JonLiu1993) [inline scripts/ci.baseline.txt:555]: Can we add this as `supports`  to vcpkg.json?

File: ports/kf5kio/vcpkg.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "kf5kio",
   "version": "5.89.0",
-  "port-version": 1,
+  "port-version": 2,
```
Comment (JonLiu1993) [inline ports/kf5kio/vcpkg.json:4]: @Neumann-A , please add the [license ](https://github.com/KDE/kio/tree/master/LICENSES)file in vcpkg.json.

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

File: ports/kf5kio/vcpkg.json (hunk 6 above)
Comment (Neumann-A) [reply ports/kf5kio/vcpkg.json:4]: Personally I think this should be done by @wrobelda. But I can add `"license": null` for you ;)

File: scripts/ci.baseline.txt (hunk 5 above)
Comment (Neumann-A) [reply scripts/ci.baseline.txt:555]: libxaw: No since `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` basically prohibits using `supports:` .

libxt: ICEd with the compiler are also not cases for `supports:`

**Review by JonLiu1993 — DISMISSED** [review]

File: ports/libxt/portfile.cmake [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+if(VCPKG_CROSSCOMPILING)
+    set(PATCHES cc_for_build.patch)
+endif()
+
+vcpkg_from_gitlab(
+    GITLAB_URL https://gitlab.freedesktop.org/xorg
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO lib/libxt
+    REF edd70bdfbbd16247e3d9564ca51d864f82626eb7 # 1.2.1
+    SHA512  c49876253dfd187e7d56a098d3d992157daefa2c25ee732eaae5818ee04513bedd807d2f265085db2e82c0b1821e152a88fb4998c0002f6ac57204543fe18566
```
Comment (Neumann-A) [inline ports/libxt/portfile.cmake:21]: Hmm a bit of testing later. Currently this only works for x86 debug due to pointer type issues. Moving this up to latest commit and adding a small patch to enable XIntPtr and XUIntPtr to be correctly pointing to intptr_t and uintptr_t for _MSC_VER enables x64 debug. The release version requires another patch to not try to open/test the '\0' file which seems to hang the application in release builds (debug builds work somehow.) [Tests were performed compiling xeyes; required patches and applications are commited in #21584]

**Comment by BillyONeal** [comment]
I'm not sure how to evaluate the correctness of these patches. Have they been submitted upstream?

**Comment by Neumann-A** [comment]
> I'm not sure how to evaluate the correctness of these patches. Have they been submitted upstream?

No upstream doesn't care about MSVC or native Windows at all. Leave a comment at the patch you have a problem to evaluate the correctness of.

**Comment by Neumann-A** [comment]
> Anything that is a meaningful and substantial ....

I already said I am fine with removing that patch and just configure libx11 without threadsafe init. All multi-threaded Xlib application should already be calling `XInitThreads` (assuming they read the docs ;) )

> initialization order consequences

there are none. `XInitThreads` just initializes some variables/locks itself so other functions in `main` can call xlib functions without caring to call `XInitThreads` before that. Basically the code should always look like:
```
int main() {
XInitThreads() // Only necessary If you plan to all call xlib functions from different threads
<other xlib functions call go here>
}
```

but as i said no need to argue here I just remove it ;)

<hr>

> Similarly libxmu/cl-build.patch sets a macro controlling an unknowably large amount of code, and the creates a global variable bringing up the nonlocal static initialization order fiasco.

`#define XTSTRINGDEFINES // Otherwise XtRImmediate is not a constant expression`

Is a funny one. It only controls how some purely string defines are handled from `<X11/StringDefs.h>` (same behavior in `<X11/Shell.h>`; `XTSTRINGDEFINES` is only used in those two headers. And those two headers are also automatically generated by the build of libX11)
```
#ifdef XTSTRINGDEFINES
#define XtNaccelerators "accelerators"
#define XtNallowHoriz "allowHoriz"
...
#else
XT_EXTERN_API _XtStringDefs_h_Const char XtStrings[]; // XtStrings is a char array containing the above names as a null terminated list of char arrays
#ifndef XtNaccelerators
#define XtNaccelerators ((String)&XtStrings[0])
#endif
#ifndef XtNallowHoriz
#define XtNallowHoriz ((String)&XtStrings[13])
#endif
....
#endif
```

Basically MSVC cannot use something like `((String)&XtStrings[13])` in const context if `libX11` has been build dynamically (it works if `libX11` is a static library). I can generally set `XTSTRINGDEFINES` in the header of `libX11` if that is more to your liking but I assumed explicitly switching is better to avoid duplication of strings in non const context. 

>  and the creates a global variable bringing up the non-local static initialization order fiasco.

I'll recheck if `const EditresBlock editDefault(BlockNone);` is required. I currently cannot remember why I added it.

**Comment by Neumann-A** [comment]
@JonLiu1993 Remove `author_response` tag?

**Comment by BillyONeal** [comment]
@JavierMatosD @valeriaconde @ras0219-msft @vicroms and I looked at this today and now that a lot of the patching has been removed agree it might be OK after some investigation of what that set macro does. Retagging reviewed.

File: ports/libxt/portfile.cmake [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+if(VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_TARGET_IS_MINGW)
+    z_vcpkg_get_cmake_vars(cmake_vars_file)
+    include("${cmake_vars_file}")
+    if(VCPKG_DETECTED_CMAKE_C_COMPILER_ID STREQUAL "MSVC")
+        vcpkg_find_acquire_program(CLANG)
+        cmake_path(GET CLANG PARENT_PATH CLANG_PARENT_PATH)
+        set(CLANG_CL "${CLANG_PARENT_PATH}/clang-cl.exe")
+        file(READ "${cmake_vars_file}" contents)
+        string(APPEND contents "\nset(VCPKG_DETECTED_CMAKE_C_COMPILER \"${CLANG_CL}\")")
+        string(APPEND contents "\nset(VCPKG_DETECTED_CMAKE_CXX_COMPILER \"${CLANG_CL}\")")
```
Comment (dg0yt) [inline ports/libxt/portfile.cmake:48]: Ouch. This port switches the compiler away from the triplet's/user's choice of MSVC to clang-cl. 

This port is an expected CI fail which unexpectedly passes now.
Ripping out this code should restore the expected fail.

Comment (BillyONeal) [reply ports/libxt/portfile.cmake:48]: Damn, I guess we missed that :(
